compiler/test/dotty/tools/dotc/semanticdb/SemanticdbTests.scala

```diff
@@ -0,0 +1,159 @@
+package dotty.tools.dotc.semanticdb
+
+import java.net.URLClassLoader
+import java.util.regex.Pattern
+import java.io.File
+import java.nio.file._
+import java.nio.charset.StandardCharsets
+import java.util.stream.Collectors
+import java.util.Comparator
+import scala.util.control.NonFatal
+import scala.collection.mutable
+import scala.jdk.CollectionConverters._
+
+import org.junit.Assert._
+import org.junit.Test
+import org.junit.experimental.categories.Category
+
+import dotty.BootstrappedOnlyTests
+import dotty.tools.dotc.Main
+import dotty.tools.dotc.semanticdb.DiffAssertions._
+import dotty.tools.dotc.semanticdb.Scala3.given
+import dotty.tools.dotc.util.SourceFile
+
+@main def updateExpect =
+  SemanticdbTests().runExpectTest(updateExpectFiles = true)
+
+@Category(Array(classOf[BootstrappedOnlyTests]))
+class SemanticdbTests with
+  val scalaFile = FileSystems.getDefault.getPathMatcher("glob:**.scala")
+  val expectFile = FileSystems.getDefault.getPathMatcher("glob:**.expect.scala")
+  // val semanticdbFile = FileSystems.getDefault.getPathMatcher("glob:**.scala.semanticdb")
+  val rootSrc = Paths.get(System.getProperty("dotty.tools.dotc.semanticdb.test"))
+  val expectSrc = rootSrc.resolve("expect")
+  val metacExpectFile = rootSrc.resolve("metac.expect")
+
+  @Category(Array(classOf[dotty.SlowTests]))
+  @Test def expectTests: Unit = runExpectTest(updateExpectFiles = false)
+
+  def runExpectTest(updateExpectFiles: Boolean): Unit =
+    val target = generateSemanticdb()
+    val errors = mutable.ArrayBuffer.empty[Path]
+    given metacSb: StringBuilder = StringBuilder(5000)
+    for source <- inputFiles().sorted do
+      val filename = source.getFileName.toString
+      val relpath = expectSrc.relativize(source)
+      val semanticdbPath = target
+        .resolve("META-INF")
+        .resolve("semanticdb")
+        .resolve(relpath)
+        .resolveSibling(filename + ".semanticdb")
+      val expectPath = source.resolveSibling(filename.replaceAllLiterally(".scala", ".expect.scala"))
+      val doc = Tools.loadTextDocument(source, relpath, semanticdbPath)
+      Tools.metac(doc, rootSrc.relativize(source))(given metacSb)
+      val obtained = trimTrailingWhitespace(SemanticdbTests.printTextDocument(doc))
+      if updateExpectFiles
```

Comment:
This code looks similar for individual `expect` files and the `metac` file below – can we try to DRY it